Added automatic event sharding

jroper opened the pull request:
This modifies event sharding so that the user can return an AggregateEventShards class, that Lagom will then use to generate a tag from the persistence ID.  This means the user doesn't have to implement the aggregateTag method on each event sub class, and doesn't necessarily have to include the entity id on each event.

This does break binary compatibility with Lagom 1.1, since `AggregateEvent.aggregateTag` now returns a supertype of `AggregateEventTag`. In practice, this shoudln't cause a problem, since the intended end user use of this method is only to implement it, it's really only ever invoked by Lagom, and returning a more specific type from a sub class is binary compatible.

jroper commented:
@patriknw this is what we were talking about.  Note that I had to get rid of the `AggregateEventTagger`, since it doesn't have access to the persistence ID, instead I now tag events in the actor.  The effect of this I guess is that if users wan't to add their own event transformers, they'll have to handle events wrapped in `Tagged` envelopes.

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/javadsl/persistence/AggregateEventTag.scala:
```diff
@@ -24,29 +24,30 @@ object AggregateEventTag {
    */
-  def shard[Event <: AggregateEvent[Event]](eventType: Class[Event], numShards: Int, entityId: String): AggregateEventTag[Event] =
-    shard(eventType, eventType.getName, numShards, entityId)
+  def sharded[Event <: AggregateEvent[Event]](eventType: Class[Event], numShards: Int): AggregateEventShards[Event] =
```
shall we remove this overload to not encourage usage of class name?

jroper replied:
If we do that, we should also deprecate the equivalent `of` method and remove the notes in the docs about using the class name.

jroper replied:
Ok, I think we should probably address this in a different issue.

jroper commented:
@patriknw I pushed a change that simplifies returning the tags from the read side processor a little, previously you returned a list of `AggregateEventTagger`, which meant you could either return a list of shards or a list of tags, and Lagom would handle that accordingly.  The implementation therefore had to be:

```java
public PSequence<AggregateEventTagger<MyEvent>> aggregateTags() {
  return TreePVector.singleton(MyEvent.TAG);
}
```

Now I've changed it so that you return a list of `AggregateEventTag`, and added an `allTags` method to `AggregateEventShards`, which means you just do this:

```java
public PSequence<AggregateEventTagger<MyEvent>> aggregateTags() {
  return MyEvent.TAG.allTags();
}
```
